design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,367 @@
+| 2    | *Outside trigger,* register to 3rd party pub/sub message handler to be notified of secrets change.<br />Ex: <br />- ZMQ (no agent)<br />- RabbitMQ |                                                              | - Do the job only when really needed - very efficient<br />- Immediate update of K8s Secret on secret change in Conjur<br />- Integrated option for registration only on relevant changes | - Need development in Conjur side to update the 3rd party component with changes<br />- Requires another component to deploy and maintain (if not ZMQ) |
+| 3    | *Outside trigger,* retrieve secrets when notified from Follower by webhook registered API | *Concern*<br />Follower does not save state and cannot save the webhook URL | - Efficient because only do the job when really needed<br />- Immediate update of K8s Secret on secret change in Conjur | - Need development in Conjur side to support notification mechanism to notify only necessary secrets providers<br />- Follower may not be able to reach secrets provider (need to save URL in Master Database)<br />- Significant development required because it requires a new event mechanism on server |
+| 4    | *Outside trigger,* retrieve secrets when notified by websocket connection | Websocket connection is opened and when there is a change on the server, it will send update | - Efficient because only do the job when really needed<br />- Immediate update of K8s Secret on secret change in Conjur | - Need development in Conjur side to support notification mechanism to notify only necessary secrets providers<br />- Very expensive to hold a socket open just for few rare notifications<br />- Significant development required because it requires a new event mechanism on server |
+
+
+
+Decision: **Solution #2, Outside trigger using pub/sub message
+**
+
+The Pub/Sub pattern is a pattern where senders publish messages without knowing who is subscribing to those messages. Once an event is emitted or a message is published, all receivers that subscribe to that event will get a notification. For our purposes, we propose to use a 3rd party pub/sub message handlers, ZMQ, the publisher, that will trigger an event whenever a Conjur secret has been updated. The Secrets Provider, the subscriber, will listen on a port for these events/messages. Once a message is received this will prompt the Secrets Provider to update the K8s Secrets with the updated Conjur Secrets.
+
+We chose to use the pub/sub pattern over waiting for a configurable amount of time because this delivers a nicer customer experience. The customer's K8s Secrets will be updated immediately which in turn avoids situations where customer applications fail to connect to their endpoints. 
```

Comment:
Also, it is allows expanding for more usecases in the future, and it is more robust as it can be monitored and audited.

Reply:
Can you add link to ZMQ? I googled but didn't find anything on it

Can you elaborate on how you protect malicious parties from putting data into the queue?